Organized functions into files and attempted to improve readability/c…

radiantdinosaurs opened the pull request:
…ontrol flow.

Created modules in the utils folder. Changed some functions to return promises (handleWriteFileRequest in write-file and classifyImage in run-tensorflow), and moved some parts of functions into small, more readable functions.

radiantdinosaurs committed:
Organized functions into files and attempted to improve readability/control flow.

Created modules in the utils folder. Changed some functions to return promises (handleWriteFileRequest in write-file and classifyImage in run-tensorflow), and moved some parts of functions into small, more readable functions.

andrewvora commented on app/app.js:
```diff
@@ -3,177 +3,49 @@
+    const base64 = request.body.base64
+    if (base64) {
+        // first, write the file to disk
+        writeFile.writeFileToDisk(base64).then((fileName) => {
```
In Node, it's typical for callback functions to be in the form `function(error, result)`. This gives consumers the ability to deal with errors. Since you probably want to bubble up errors to the route handling logic, you'll want to be able to receive errors from any function calls here

andrewvora replied:
Disregard, didn't notice this was a promise

andrewvora commented on app/app.js:
```diff
@@ -3,177 +3,49 @@
+            // next, run TensorFlow's trained model
+            runTensorFlow.classifyImage(fileName).then((tensorFlowResult) => {
+                // finally, send the result
+                if (tensorFlowResult instanceof Error) {
```
This works, but it's not really node convention. See the above comment for how errors are normally handled in Node

EDIT: In this case, you just want to throw the error if there is one, since that will be handled by your catch

andrewvora replied:
This comment still applies though

andrewvora commented on app/app.js:
```diff
@@ -3,177 +3,49 @@
+            }).catch(error => {
+                console.log(error)
+                response.status(500).send({status: 500,
+                    message: 'Internal error encountered while running TensorFlow script.'})
```
Since this is the public facing response, you don't want to reveal what your stack consists of. Consumers shouldn't even know this is a node project

andrewvora commented on app/utils/parse-base64.js:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Validates a file format by analyzing the data label (the first 15 characters of a base64 string)
+ * @param {string} base64 - base64 string
+ * @returns {string|Error} - validated file format or Error object
```
See comment above about explicitly returning an error. If you have to return error information, do it as an object literal or a class. Something that's well-defined like guaranteeing there's a `msg` property

andrewvora commented on app/utils/parse-base64.js:
```diff
@@ -0,0 +1,32 @@
+ */
+const validateFileFormat = (base64) => {
+    if (base64) {
+        if (base64.substring(0, 15).includes('jpeg')) return 'jpeg'
```
JPEG files can have the extension JPG, too

andrewvora commented on app/utils/parse-base64.js:
```diff
@@ -0,0 +1,32 @@
+    } else {
+        let error = new Error('Cannot validate file format if argument is undefined')
+        error.code = 400
+        return error
```
Yeah, this would work totally fine as an object literal or a class instance. Just don't return `Error`s

andrewvora commented on app/utils/parse-base64.js:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Strips data label from a base64 string
+ * @param base64 - base64 string
+ * @returns {string|Error} - stripped base64 or Error object
```
Misleading documentation

andrewvora commented on app/utils/parse-base64.js:
```diff
@@ -0,0 +1,32 @@
+ * @returns {string|Error} - stripped base64 or Error object
+ */
+const stripDataLabel = (base64) => {
+    return base64.replace(/^data:image\/jpeg;base64,/, '')
```
Should there be null-handling here?

andrewvora commented on app/utils/tensorflow/parse-tensorflow.js:
```diff
@@ -0,0 +1,37 @@
+    } else {
+        let error = new Error('Cannot parse undefined argument.')
+        error.code = 400
+        return error
```
Same comment here about returning Errors

andrewvora replied:
You could tone down the number of comments. Some of it is pretty explicit

andrewvora commented on app/utils/write-file.js:
```diff
@@ -0,0 +1,83 @@
+        const fileFormat = parseBase64.validateFileFormat(base64)
+        if (fileFormat instanceof Error) return fileFormat
+        else {
+            if (fileFormat.includes('jpeg')) return writeJpegToDisk(base64, fileName)
```
Do you need handling for `jpg`?

andrewvora commented on app/utils/write-file.js:
```diff
@@ -0,0 +1,83 @@
+    } else {
+        let error = new Error('Cannot write file to disk if arguments are undefined')
+        error.code = 400
+        return error
```
Same comment about returning errors

andrewvora reviewed (changes requested):
It's looking good, but needs a few changes before it can be merged in. While you're looking over the code again, look for unit testable code and write some unit tests

radiantdinosaurs committed:
Organized file structure, created some unit tests, and fixed error handling.

Divided utils into proper folders and moved function validateFileFormat from parse-base64 into write-file. Created a file for unit tests where appropriate. Added in return-error to handle creating Errors, and fixed error handling across almost all files.

andrewvora reviewed (approved):
Things are in a better state than they were before. There's some small changes that can be made, but we can worry about that as this application gets larger. For now, we can get this merged in and refactor later
